common/acl/camanager.go

```diff
@@ -0,0 +1,303 @@
+	}
+	hash, newHash, ok := ValidateCACerts(cacert, tokenParts[0])
+	if !ok {
+		return fmt.Errorf("failed to validate CA certificate. tokenCAhash: %s, CAhash: %s", hash, newHash)
+	}
+
+	// save the ca.crt to file
+	ca, err := x509.ParseCertificate(cacert)
+	if err != nil {
+		return fmt.Errorf("failed to parse the CA certificate, error: %v", err)
+	}
+
+	if err = certutil.WriteCert(m.caFile, ca); err != nil {
+		return fmt.Errorf("failed to save the CA certificate to file: %s, error: %v", m.caFile, err)
+	}
+
+	// get the edge.crt
+	caPem := pem.EncodeToMemory(&pem.Block{Bytes: cacert, Type: cert.CertificateBlockType})
+	pk, edgeCert, err := m.GetNodeCert(m.certURL, caPem, tls.Certificate{}, strings.Join(tokenParts[1:], "."))
+	if err != nil {
+		return fmt.Errorf("failed to get edge certificate from the cloudcore, error: %v", err)
+	}
+
+	// save the edge.crt to the file
+	crt, _ := x509.ParseCertificate(edgeCert)
+	if err = certutil.WriteKeyAndCert(m.keyFile, m.certFile, pk, crt); err != nil {
+		return fmt.Errorf("failed to save the edge key and certificate to file: %s, error: %v", m.certFile, err)
+	}
+
+	return nil
+}
+
+// generateKey realizes the acl application by token
+func (m *caManager) generateKey() error {
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return fmt.Errorf("failed to generate key, error: %v", err)
+	}
+	// save the private key to the file
+	if err = certutil.WriteKey(m.keyFile, pk); err != nil {
+		return fmt.Errorf("failed to save the private key %s, error: %v", m.keyFile, err)
+	}
+	return nil
+}
+
+// getCA returns the CA in pem format.
+func (m *caManager) getCA() ([]byte, error) {
+	return ioutil.ReadFile(m.caFile)
+}
+
+// GetNodeCert applies for the certificate from cloudcore
+func (m *caManager) GetNodeCert(url string, capem []byte, cert tls.Certificate, token string) (*ecdsa.PrivateKey, []byte, error) {
+	pk, csr, err := m.getCSR()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create CSR: %v", err)
+	}
+
+	client, err := http.NewHTTPClientWithCA(capem, cert)
+	if err != nil {
+		return nil, nil, fmt.Errorf("falied to create http client:%v", err)
+	}
+
+	nodeName, isExist := os.LookupEnv(meshConstants.MY_NODE_NAME)
+	if !isExist {
+		klog.Errorf("env %s not exist", meshConstants.MY_NODE_NAME)
+	}
+
+	usages := []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth}
+	req, err := buildRequest(nethttp.MethodGet, url, bytes.NewReader(csr), token, nodeName, usages)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate http request:%v", err)
+	}
+
+	var res *nethttp.Response
+	for i := 0; i < retry; i++ {
+		res, err = http.SendRequest(req, client)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	defer res.Body.Close()
+
+	content, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+	if res.StatusCode != 200 {
+		return nil, nil, fmt.Errorf(string(content))
```

Comment:
just format the response body is not bad:
1. when status code != 200, probably empty content
2. content may have '%s' substring
3. content may have huge size

So maybe something likes:
```suggestion
		return nil, nil, fmt.Errorf("status code: %d, content: %s", res.StatusCode, string(content)[:50])
```

Reply:
fix. If the size of content is bigger than 50, I will just use string(content)[:50]